[TECH 425] GeoIP Validators Info Service

jessicaqgomez opened the pull request:
**Description:**
**1.** A new endpoint was added to unify the information of the validators list (connected and disconnected) with their geographical location based on their ip.
To obtain this information we are querying the endpoint getCurrentValidators and getPeers, and unify the data of both in this single endpoint.
**E.g.**
**Validator Disconnected**
{
  "name": "GeoIPInfo",
  "value": [
    {
      "nodeID": "NodeID-xxxxxxxx",
      "IP": "",
      "txID": "xxxxxxxx",
      "connected": false,
      "startTime": "2022-12-05 04:04:20",
      "endTime": "2022-12-28 04:12:14",
      "duration": "23 Days",
      "uptime": "0.0152",
      "country": "",
      "lng": 0,
      "lat": 0,
      "countryISO": "",
      "city": ""
    }]
   }
**Validator Connected**
{
  "name": "GeoIPInfo",
  "value": [
       {
          "nodeID": "NodeID-xxxxx",
          "IP": "xxx.xxx.xxx.xxx:xxxx",
          "txID": "xxxxxxx",
          "connected": true,
          "startTime": "2022-08-24 07:52:51",
          "endTime": "2022-12-31 09:02:12",
          "duration": "129 Days",
          "uptime": "0.9999",
          "country": "Germany",
          "lng": 12.365,
          "lat": 50.475,
          "countryISO": "DE",
          "city": "Falkenstein"
   }]
}

**2.** For the configuration of this endpoint, a new option was added in the services section of the config file, in which the url of the GeoIp service and its respective token must be specified. 
...
   "services": {
    "db": {
      "dsn": "",
      "driver": ""
    },
    "geoIP": {
      "urlEndpoint":"http://ip-api.com/json/",
      "autorizationToken": ""
    }

jessicaqgomez committed:
GeoIP Validators Info Service

jessicaqgomez committed:
GeoIP Validators Info Service

jessicaqgomez committed:
correct lint errors

jessicaqgomez committed:
correct lint errors

jessicaqgomez committed:
Lint security alert corrections

jessicaqgomez committed:
Lint security alert corrections

jessicaqgomez committed:
Lint security alert corrections

jessicaqgomez committed:
Lint security alert corrections

Chriserus reviewed (commented):
LGTM, but keep in mind I'm not yet proficient in GoLang and you should get a review from a more competent GoLang dev! Also, I see you requested review from @Ayagoumi and @aeddaqqa, they are mainly frontend developers and to my knowledge they do not know GO.

jessicaqgomez commented:
you're right, there was a confusion with the reviewers.
Thanks for letting me know

c4t-ag commented on cfg/cfg.go:
```diff
@@ -104,6 +112,80 @@ type Filter struct {
 	Min uint32 `json:"min"`
 	Max uint32 `json:"max"`
 }
+type IPAPIResponse struct {
```
This and the following lines are definitions which are IMO not related to config.
This is also true for the Aggrgate interface / struct definitions above, not sure if they came in later or are original ava code.
Please find a better place for defining them

jessicaqgomez replied:
Ok, changed these structures from cfg.go to collections.go

jessicaqgomez committed:
move geoIP structs from cfg.go to collections.go file

jessicaqgomez committed:
move geoIP structs from cfg.go to collections.go file

jessicaqgomez committed:
read env variable from system

jessicaqgomez committed:
read env variable from system

c4t-ag reviewed (approved):
Nice, thx!
Pls. PR against dev branch instead c4t @jessicaqgomez!

c4t-ag reviewed (changes requested):
Wrong branch fro PR, use dev pls.

jessicaqgomez commented:
@c4t-ag ok, I already changed the branch PR

charalarg commented on utils/validator_info.go:
```diff
@@ -0,0 +1,233 @@
+	return strings.Split(dateFTime.String(), " -")[0]
+}
+func getDuration(startTime string, endTime string) string {
+	start, _ := time.Parse("2006-01-02 15:04:05", startTime)
```
I think i don't get why we use these hardcoded dates in order to get the difference. Could you please explain?

jessicaqgomez replied:
Time.Parse  parses a formatted string and returns the time value it represents. 
`func Parse(layout, value string) (Time, error)`
These are predefined layouts for use in Time.Format and time.Parse, each element shows by example the formatting of an element of the reference time. Only these values are recognized: 
	**Year:** "2006" "06"
	**Month:** "Jan" "January" "01" "1"
        Day of the week: "Mon" "Monday"
	**Day of the month:** "2" "_2" "02"
	Day of the year: "__2" "002"
	**Hour:** "15" "3" "03" (PM or AM)
	**Minute:** "4" "04"
	**Second:** "5" "05"
	AM/PM mark: "PM"

`https://go.dev/src/time/format.go`
`https://pkg.go.dev/time#Parse`

The startTime and endTime params in duration function are strings with this format ("yyyy-mm-dd hh:mm:ss"), we need to have an time variable to use the sub function.

charalarg replied:
Didn't know about these dates.. Wouldn't be better to use a constant instead of the hardcoded string `2006-01-02 15:04:05` as layout? For example smthg like `time.UnixDate` instead? in order to avoid confusion

charalarg commented on utils/validator_info.go:
```diff
@@ -0,0 +1,233 @@
+	var info models.Validator
+	if peerFlag {
+		geoIPInfo := GetLocationByIP(peer.IP, config)
+		info = models.Validator{
```
I would suggest to avoid duplicate code for creating the `Validator` obj. IMO the obj could be initialized ones and if `peerFlag` then the rest of the values could be filled.

charalarg commented on utils/validator_info.go:
```diff
@@ -0,0 +1,233 @@
+	}
+	return info
+}
+func GetCurrentValidators(rpc string) models.ValidatorsResponse {
```
I would suggest to define the request params (endpoint, payload, headers, etc.) somewhere out of these functions instead of being hardcoded here, in order to be more accessible for changes afterwards.

charalarg commented on utils/validator_info.go:
```diff
@@ -0,0 +1,233 @@
+	return response
+}
+
+func GetPeers(rpc string) models.PeersResponse {
```
Same here, I would suggest to define the request params (endpoint, payload, headers, etc.) somewhere out of these functions instead of being hardcoded here, in order to be more accessible for changes afterwards.

charalarg commented on utils/validator_info.go:
```diff
@@ -0,0 +1,233 @@
+	req, err := http.NewRequest("POST", url, payload)
+
+	if err != nil {
+		fmt.Println(err)
```
I propose to use the logging logic used in the rest of the project for consistency. Same for the other error printing lines bellow.

jessicaqgomez committed:
add recomendations about error control, structs and functions

jessicaqgomez committed:
add recomendations about error control, structs and functions

jessicaqgomez commented:
@charalarg, I added the recommendations you requested in the last commit.

charalarg commented on utils/validator_info.go:
```diff
@@ -22,179 +24,184 @@ func PeerIndex(peers *models.PeersResponse, nodeID string) int {
-	end, _ := time.Parse("2006-01-02 15:04:05", endTime)
+
+func getDuration(startTime string, endTime string) (string, error) {
+	start, err := time.Parse(timeLayout, startTime)
```
Maybe I didn't describe well my thought. Instead of defining `timeLayout` why don't we use in this line the `time.UnixDate`? This date is already defined in the `time` package in the `UnixDate` varibale and we can directly use it instead of defining it again.

jessicaqgomez committed:
Change from time.parse to time.Unix in duration Function

jessicaqgomez committed:
Change from time.parse to time.Unix in duration Function

jessicaqgomez commented:
@c4t-ag, i already change the branch in the PR to dev

c4t-ag commented on utils/validator_info.go:
```diff
@@ -0,0 +1,240 @@
+		return *geoValidatorsInfo, err
+	}
+	peers, _ := GetPeers(rpc)
+	if !reflect.DeepEqual(validators, models.ValidatorsResponse{}) && !reflect.DeepEqual(peers, models.PeersResponse{}) &&
```
Suggestion:
Line 57: var validatorList []*models.Validator{}
then remove the complete huge (and expensive) if statement (beside the fact that the latter 2 conditions are not really required)

Side note: In real life (multiple nodes) you'll run into the issue that you get a validator but incomplete peers list.
Every node stores only a bunch of peers, but in general not all (only neighbors)

c4t-ag reviewed (changes requested):
The big question to answer here is: how do you want to deal with incomplete list of peers.
From concept (this is at least what I have learned so far) a node does not keep all peers, but only a bunch of them to select a few from.

jessicaqgomez committed:
Merge branch 'feature/TECH-425_GeoIp_ValidatorsInfo' of https://github.com/chain4travel/magellan into feature/TECH-425_GeoIp_ValidatorsInfo

jessicaqgomez committed:
change rpc calls to caminogo functions and delete unnecessary conditionals

jessicaqgomez committed:
Fix lint recommendations

jessicaqgomez commented:
@c4t-ag Ok, fixed in the last commit.
Additionally I added some changes: the call to the rpc is no longer made, instead the caminogo functions are used.

On the other hand, the reason why we use the peers is to obtain the ip address of the validators at the moment, since no other endpoint has this information available.
It has already been considered that the validators that are not in this list of peers do not have geoIp information but do have the remaining information.

We are considering with alex the possibility of adding an endpoint to the rpc that has this ip info but for the moment we will continue using the information provided by the peers.
